aptos-move/move-examples/data_structures/sources/smart_table.move

```diff
@@ -0,0 +1,564 @@
+/// A smart table implementation based on linear hashing. (https://en.wikipedia.org/wiki/Linear_hashing)
+/// Compare to Table, it uses less storage slots but has higher chance of collision, it's a trade-off between space and time.
+/// Compare to other implementation, linear hashing splits one bucket a time instead of doubling buckets when expanding to avoid unexpected gas cost.
+/// SmartTable uses faster hash function SipHash instead of cryptographically secure hash functions like sha3-256 since it tolerates collisions.
+module aptos_std::smart_table {
+    use std::error;
+    use std::vector;
+    use aptos_std::aptos_hash::sip_hash_from_value;
+    use aptos_std::table_with_length::{Self, TableWithLength};
+    use aptos_std::type_info::size_of_val;
+    use aptos_std::math64::max;
+
+    /// Key not found in the smart table
+    const ENOT_FOUND: u64 = 1;
+    /// Smart table capacity must be larger than 0
+    const EZERO_CAPACITY: u64 = 2;
+    /// Cannot destroy non-empty hashmap
+    const ENOT_EMPTY: u64 = 3;
+    /// Key already exists
+    const EALREADY_EXIST: u64 = 4;
+    /// Invalid load threshold percent to trigger split.
+    const EINVALID_LOAD_THRESHOLD_PERCENT: u64 = 5;
+    /// Invalid target bucket size.
+    const EINVALID_TARGET_BUCKET_SIZE: u64 = 6;
+
+    /// SmartTable entry contains both the key and value.
+    struct Entry<K, V> has copy, drop, store {
+        hash: u64,
+        key: K,
+        value: V,
+    }
+
+    struct SmartTable<K, V> has store {
+        buckets: TableWithLength<u64, vector<Entry<K, V>>>,
+        num_buckets: u64,
+        // number of bits to represent num_buckets
+        level: u8,
```

Comment:
Just double checking - this doesn't allow more than 256 (2^8) buckets. What's the rationale behind this?

Reply:
Actually it is 2^{level <= 256}.